src/dim_helpers/ConvDims.jl

```diff
@@ -6,13 +6,19 @@ export ConvDims
 Type system-level information about convolution dimensions. Critical for things like
 `im2col!()` to generate efficient code, and helpful to reduce the number of kwargs
 getting passed around.
-
-We don't want to specialize on things like image size/channel count, so we generally
-store those as fields, just for convenience, and to allow for non-breaking changes when
-we decide we _do_ want to specialize on those values.  We always want to specialize on
-things like stride, padding, dilation, and kernel flipping though.
 """
-abstract type ConvDims{N, S, P, D, F} end
+abstract type ConvDims{N} end
+
+@inline spatial_dims(::ConvDims{N}) where N = N
+@inline groupcount(c::ConvDims) = 1
+
+function input_size(::ConvDims) end
+function kernel_size(::ConvDims) end
+
+function stride(::ConvDims) end
+function padding(::ConvDims) end
+function dilation(::ConvDims) end
+function flipkernel(::ConvDims) end
```

Comment:
What are these methods for? Is nothing being used somewhere to mean "unknown"? I don't see `=== nothing` but did not look everywhere.

(It might also be less mysterious to write them `input_size(::ConvDims) = nothing` etc, if the are needed.)

Reply:
I guess it was meant to be a "fallback" (or interface indication) if someone is using `ConvDims` only.
IMO it would make sense to either remove them completely or remove `(::ConvDims)` part.

We can leave `function input_size end` in code to indicate "interface" that dims should implement as was suggested in [JuliaLang Antipatterns](https://www.oxinabox.net/2020/04/19/Julia-Antipatterns.html#notimplemented-exceptions).

Reply:
Ok. Zero-method stubs are better as documentation, `function input_size end`. Still a little bit of a puzzle, since you wonder whether they are being defined e.g. in the root module with methods added from elsewhere. IMO if you were ever going to extend this, you'd look at one of the working examples, which defines all of them together, to figure out what's needed.